mindmeld/components/nlp.py

```diff
@@ -1507,10 +1542,22 @@ def process_entity(self, query, entities, entity_index, verbose=False):
 
         if self.role_classifier.roles:
             # Only run role classifier if there are roles!
-            if verbose:
-                role = self.role_classifier.predict_proba(query, entities, entity_index)
-                entity.entity.role = role[0][0]
-                confidence_score = dict(role)
+            if verbose or allowed_nlp_classes:
+                roles = self.role_classifier.predict_proba(query, entities, entity_index)
+                for role in roles:
+                    # the role confidences are sorted, so we will always be able to pick
+                    # the highest confidence role that matches the allowed_nlp_classes
+                    role_type = role[0]
+
+                    if not allowed_nlp_classes:
+                        entity.entity.role = role_type
+                        break
+
+                    if allowed_nlp_classes and role_type in allowed_nlp_classes:
```

Comment:
unnecessary to check again
```suggestion
                    if role_type in allowed_nlp_classes:
```